File: src/test/java/svnserver/ext/gitlfs/storage/network/LfsHttpStorageTest.java
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.ext.gitlfs.storage.network;

import com.google.common.hash.Hashing;
import com.google.common.io.ByteStreams;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.NameValuePair;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.message.BasicNameValuePair;
import org.eclipse.jgit.util.Holder;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.mapdb.DBMaker;
import org.testng.Assert;
import org.testng.annotations.Test;
import org.tmatesoft.svn.core.SVNErrorMessage;
import org.tmatesoft.svn.core.SVNLock;
import org.tmatesoft.svn.core.io.ISVNLockHandler;
import org.tmatesoft.svn.core.io.SVNRepository;
import ru.bozaro.gitlfs.client.Client;
import ru.bozaro.gitlfs.client.auth.CachedAuthProvider;
import ru.bozaro.gitlfs.client.exceptions.RequestException;
import ru.bozaro.gitlfs.common.JsonHelper;
import ru.bozaro.gitlfs.common.data.Link;
import ru.bozaro.gitlfs.common.data.Operation;
import svnserver.SvnTestHelper;
import svnserver.SvnTestServer;
import svnserver.VcsAccessEveryone;
import svnserver.VcsAccessNoAnonymous;
import svnserver.auth.LocalUserDB;
import svnserver.auth.User;
import svnserver.auth.UserDB;
import svnserver.config.SharedConfig;
import svnserver.context.LocalContext;
import svnserver.context.SharedContext;
import svnserver.ext.gitlfs.server.LfsServer;
import svnserver.ext.gitlfs.storage.LfsReader;
import svnserver.ext.gitlfs.storage.LfsStorage;
import svnserver.ext.gitlfs.storage.LfsStorageFactory;
import svnserver.ext.gitlfs.storage.LfsWriter;
import svnserver.ext.gitlfs.storage.local.LfsLocalStorageTest;
import svnserver.ext.gitlfs.storage.memory.LfsMemoryStorage;
import svnserver.ext.web.config.WebServerConfig;
import svnserver.ext.web.server.WebServer;
import svnserver.repository.VcsAccess;

import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static svnserver.server.SvnFilePropertyTest.propsBinary;

/**
 * Simple test for LfsLocalStorage.
 *
 * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
 * @author Marat Radchenko <m.radchenko@corp.mail.ru>
 */
public final class LfsHttpStorageTest {

  @Test
  public void commitToRemoteLfs() throws Exception {
    // Create users
    final LocalUserDB users = new LocalUserDB();
    final User user = users.add(SvnTestServer.USER_NAME, "test", "Test User", "test@example.com");
    Assert.assertNotNull(user);
    // Create shared context
    try (SharedContext sharedContext = SharedContext.create(Paths.get("/nonexistent"), "realm", DBMaker.memoryDB().make(), Collections.singletonList(new WebServerConfig(0)))) {
      final WebServer webServer = sharedContext.sure(WebServer.class);
      sharedContext.add(LfsServer.class, new LfsServer("t0ken", 0, 0));
      sharedContext.add(UserDB.class, users);
      sharedContext.ready();
      // Create local context
      LocalContext localContext = new LocalContext(sharedContext, "example");

      localContext.add(VcsAccess.class, new VcsAccessEveryone());

      final LfsMemoryStorage backendStorage = new LfsMemoryStorage();
      localContext.add(LfsStorage.class, backendStorage);
      // Register storage
      sharedContext.sure(LfsServer.class).register(localContext, localContext.sure(LfsStorage.class));

      final URI url = webServer.getBaseUrl().resolve("example.git/").resolve(LfsServer.SERVLET_AUTH);

      try (SvnTestServer server = SvnTestServer.createEmpty(null, false, SvnTestServer.LfsMode.None, new GitAsSvnLfsHttpStorage(url, user))) {
        final SVNRepository svnRepository = server.openSvnRepository();
        SvnTestHelper.createFile(svnRepository, ".gitattributes", "* -text\n*.txt filter=lfs diff=lfs merge=lfs -text", propsBinary);

        final byte[] data = LfsLocalStorageTest.bigFile();

        SvnTestHelper.createFile(svnRepository, "1.txt", data, propsBinary);
        SvnTestHelper.checkFileContent(svnRepository, "1.txt", data);

        Assert.assertEquals(backendStorage.getFiles().size(), 1);
        final byte[] lfsBytes = backendStorage.getFiles().values().iterator().next();
        Assert.assertNotNull(lfsBytes);
        Assert.assertEquals(lfsBytes, data);

        final Holder<SVNLock> lockHolder = new Holder<>(null);
        svnRepository.lock(Collections.singletonMap("1.txt", svnRepository.getLatestRevision()), null, false, new ISVNLockHandler() {
          @Override
          public void handleLock(String path, SVNLock lock, SVNErrorMessage error) {
            Assert.assertEquals(path, "/1.txt");
            lockHolder.set(lock);
          }

          @Override
          public void handleUnlock(String path, SVNLock lock, SVNErrorMessage error) {
            Assert.fail();
          }
        });
        final SVNLock lock = lockHolder.get();
        Assert.assertNotNull(lock);
        Assert.assertEquals(lock.getPath(), "/1.txt");
        Assert.assertNotNull(lock.getID());
        Assert.assertEquals(lock.getOwner(), SvnTestServer.USER_NAME);
      }
    }
  }

  @Test
  public void server() throws Exception {
    // Create users
    final LocalUserDB users = new LocalUserDB();
    final User user = users.add("test", "test", "Test User", "test@example.com");
    Assert.assertNotNull(user);
    // Create shared context
    try (SharedContext sharedContext = SharedContext.create(Paths.get("/nonexistent"), "realm", DBMaker.memoryDB().make(), Collections.singletonList(new WebServerConfig(0)))) {
      final WebServer webServer = sharedContext.sure(WebServer.class);
      sharedContext.add(LfsServer.class, new LfsServer("t0ken", 0, 0));
      sharedContext.add(UserDB.class, users);
      sharedContext.ready();
      // Create local context
      LocalContext localContext = new LocalContext(sharedContext, "example");
      localContext.add(VcsAccess.class, new VcsAccessNoAnonymous());
      localContext.add(LfsStorage.class, new LfsMemoryStorage());
      // Register storage
      sharedContext.sure(LfsServer.class).register(localContext, localContext.sure(LfsStorage.class));

      final byte[] data = LfsLocalStorageTest.bigFile();

      final String oid = "sha256:" + Hashing.sha256().hashBytes(data).toString();

      final URI url = webServer.getBaseUrl().resolve("example.git/").resolve(LfsServer.SERVLET_AUTH);
      final LfsHttpStorage storage = new GitAsSvnLfsHttpStorage(url, user);

      // Check file is not exists
      Assert.assertNull(storage.getReader(oid, -1));

      // Write new file
      try (final LfsWriter writer = storage.getWriter(user)) {
        writer.write(data);
        Assert.assertEquals(writer.finish(null), oid);
      }

      // Write new file AGAIN
      try (final LfsWriter writer = storage.getWriter(user)) {
        writer.write(data);
        Assert.assertEquals(writer.finish(null), oid);
      }

      // Read old file.
      final LfsReader reader = storage.getReader(oid, -1);
      Assert.assertNotNull(reader);
      Assert.assertNull(reader.getMd5());
      Assert.assertEquals(reader.getSize(), data.length);

      try (final InputStream stream = reader.openStream()) {
        final byte[] actual = ByteStreams.toByteArray(stream);
        Assert.assertEquals(actual, data);
      }

      LfsLocalStorageTest.checkLfs(storage, user);
      LfsLocalStorageTest.checkLfs(storage, user);

      LfsLocalStorageTest.checkLocks(storage, user);
    }
  }

  private static final class GitAsSvnLfsHttpStorage extends LfsHttpStorage implements LfsStorageFactory, SharedConfig {
    @NotNull
    private final URI authUrl;
    @NotNull
    private final User user;

    private GitAsSvnLfsHttpStorage(@NotNull URI authUrl, @NotNull User user) {
      this.authUrl = authUrl;
      this.user = user;
    }

    @Override
    public @NotNull LfsStorage createStorage(@NotNull LocalContext context) {
      return this;
    }

    @Override
    protected @NotNull Client lfsClient(@NotNull User unused) {
      final CloseableHttpClient httpClient = LfsHttpStorage.createHttpClient();

      final CachedAuthProvider authProvider = new CachedAuthProvider() {
        @Override
        protected @NotNull Link getAuthUncached(@NotNull Operation operation) throws IOException {
          final HttpPost post = new HttpPost(authUrl);
          final List<NameValuePair> params = new ArrayList<>();
          addParameter(params, "secretToken", "t0ken");

          if (user.isAnonymous()) {
            addParameter(params, "mode", "anonymous");
          } else {
            addParameter(params, "mode", "username");
            addParameter(params, "userId", user.getUserName());
          }

          post.setEntity(new UrlEncodedFormEntity(params));
          try {
            final HttpResponse response = httpClient.execute(post);
            if (response.getStatusLine().getStatusCode() == HttpStatus.SC_OK) {
              return JsonHelper.mapper.readValue(response.getEntity().getContent(), Link.class);
            }
            throw new RequestException(post, response);
          } finally {
            post.abort();
          }
        }

        private void addParameter(@NotNull List<NameValuePair> params, @NotNull String key, @Nullable String value) {
          if (value != null) {
            params.add(new BasicNameValuePair(key, value));
          }
        }
      };
      return new Client(authProvider, httpClient);
    }

    @Override
    public void create(@NotNull SharedContext context) {
      context.add(LfsStorageFactory.class, this);
    }

    @Override
    public void close() {
    }
  }

}

File: src/test/kotlin/svnserver/ext/gitlfs/storage/network/LfsHttpStorageTest.kt
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.ext.gitlfs.storage.network

import com.google.common.hash.Hashing
import com.google.common.io.ByteStreams
import org.apache.http.HttpResponse
import org.apache.http.HttpStatus
import org.apache.http.NameValuePair
import org.apache.http.client.entity.UrlEncodedFormEntity
import org.apache.http.client.methods.HttpPost
import org.apache.http.message.BasicNameValuePair
import org.eclipse.jgit.util.Holder
import org.mapdb.DBMaker.memoryDB
import org.testng.Assert
import org.testng.annotations.Test
import org.tmatesoft.svn.core.SVNErrorCode
import org.tmatesoft.svn.core.SVNErrorMessage
import org.tmatesoft.svn.core.SVNException
import org.tmatesoft.svn.core.SVNLock
import org.tmatesoft.svn.core.io.ISVNLockHandler
import org.tmatesoft.svn.core.io.SVNRepository
import ru.bozaro.gitlfs.client.Client
import ru.bozaro.gitlfs.client.auth.CachedAuthProvider
import ru.bozaro.gitlfs.client.exceptions.RequestException
import ru.bozaro.gitlfs.common.JsonHelper
import ru.bozaro.gitlfs.common.data.Link
import ru.bozaro.gitlfs.common.data.Operation
import svnserver.SvnTestHelper
import svnserver.SvnTestServer
import svnserver.VcsAccessEveryone
import svnserver.VcsAccessNoAnonymous
import svnserver.auth.LocalUserDB
import svnserver.auth.User
import svnserver.auth.UserDB
import svnserver.config.SharedConfig
import svnserver.context.LocalContext
import svnserver.context.SharedContext
import svnserver.ext.gitlfs.server.LfsServer
import svnserver.ext.gitlfs.storage.LfsStorage
import svnserver.ext.gitlfs.storage.LfsStorageFactory
import svnserver.ext.gitlfs.storage.local.LfsLocalStorageTest
import svnserver.ext.gitlfs.storage.memory.LfsMemoryStorage
import svnserver.ext.web.config.WebServerConfig
import svnserver.ext.web.server.WebServer
import svnserver.repository.VcsAccess
import svnserver.server.SvnFilePropertyTest
import java.io.IOException
import java.net.URI
import java.nio.file.Paths
import java.util.*

/**
 * Simple test for LfsLocalStorage.
 *
 * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
 * @author Marat Radchenko <marat@slonopotamus.org>
 */
class LfsHttpStorageTest {
    @Test
    @Throws(Exception::class)
    fun commitToRemoteLfs() {
        // Create users
        val users = LocalUserDB()
        val user = users.add(SvnTestServer.USER_NAME, "test", "Test User", "test@example.com")
        Assert.assertNotNull(user)
        SharedContext.create(Paths.get("/nonexistent"), "realm", memoryDB().make(), listOf(WebServerConfig(0))).use { sharedContext ->
            val webServer = sharedContext.sure(WebServer::class.java)
            sharedContext.add(LfsServer::class.java, LfsServer("t0ken", 0, 0F))
            sharedContext.add(UserDB::class.java, users)
            sharedContext.ready()
            // Create local context
            val localContext = LocalContext(sharedContext, "example")
            localContext.add(VcsAccess::class.java, VcsAccessEveryone())
            val backendStorage = LfsMemoryStorage()
            localContext.add(LfsStorage::class.java, backendStorage)
            // Register storage
            sharedContext.sure(LfsServer::class.java).register(localContext, localContext.sure(LfsStorage::class.java))
            val url = webServer.getBaseUrl().resolve("example.git/").resolve(LfsServer.SERVLET_AUTH)
            SvnTestServer.createEmpty(null, false, SvnTestServer.LfsMode.None, GitAsSvnLfsHttpStorage(url, user!!)).use { server ->
                val svnRepository: SVNRepository = server.openSvnRepository()
                SvnTestHelper.createFile(svnRepository, ".gitattributes", "* -text\n*.txt filter=lfs diff=lfs merge=lfs -text", SvnFilePropertyTest.propsBinary)
                val data: ByteArray = LfsLocalStorageTest.bigFile()
                SvnTestHelper.createFile(svnRepository, "1.txt", data, SvnFilePropertyTest.propsBinary)
                SvnTestHelper.checkFileContent(svnRepository, "1.txt", data)
                Assert.assertEquals(backendStorage.files.size, 1)
                val lfsBytes = backendStorage.files.values.iterator().next()
                Assert.assertNotNull(lfsBytes)
                Assert.assertEquals(lfsBytes, data)
                val lockHolder = Holder<SVNLock?>(null)
                svnRepository.lock(Collections.singletonMap("1.txt", svnRepository.latestRevision), null, false, object : ISVNLockHandler {
                    override fun handleLock(path: String, lock: SVNLock?, error: SVNErrorMessage?) {
                        Assert.assertEquals(path, "/1.txt")
                        lockHolder.set(lock)
                    }

                    override fun handleUnlock(path: String, lock: SVNLock?, error: SVNErrorMessage?) {
                        Assert.fail()
                    }
                })
                val lock = lockHolder.get()
                Assert.assertNotNull(lock)
                Assert.assertEquals(lock!!.path, "/1.txt")
                Assert.assertNotNull(lock.id)
                Assert.assertEquals(lock.owner, SvnTestServer.USER_NAME)
                Assert.assertEquals(backendStorage.locks.size, 1)
                try {
                    SvnTestHelper.modifyFile(svnRepository, "1.txt", "blabla", -1, null)
                    Assert.fail()
                } catch (e: SVNException) {
                    Assert.assertEquals(e.errorMessage.errorCode, SVNErrorCode.FS_BAD_LOCK_TOKEN)
                }
                SvnTestHelper.modifyFile(svnRepository, "1.txt", "blabla", -1, Collections.singletonMap(lock.path, lock.id))
                Assert.assertEquals(backendStorage.locks.size, 0)
            }
        }
    }

    @Test
    @Throws(Exception::class)
    fun server() {
        // Create users
        val users = LocalUserDB()
        val user = users.add("test", "test", "Test User", "test@example.com")
        Assert.assertNotNull(user)
        SharedContext.create(Paths.get("/nonexistent"), "realm", memoryDB().make(), listOf(WebServerConfig(0))).use { sharedContext ->
            val webServer = sharedContext.sure(WebServer::class.java)
            sharedContext.add(LfsServer::class.java, LfsServer("t0ken", 0, 0F))
            sharedContext.add(UserDB::class.java, users)
            sharedContext.ready()
            // Create local context
            val localContext = LocalContext(sharedContext, "example")
            localContext.add(VcsAccess::class.java, VcsAccessNoAnonymous())
            localContext.add(LfsStorage::class.java, LfsMemoryStorage())
            // Register storage
            sharedContext.sure(LfsServer::class.java).register(localContext, localContext.sure(LfsStorage::class.java))
            val data: ByteArray = LfsLocalStorageTest.bigFile()
            val oid = "sha256:" + Hashing.sha256().hashBytes(data).toString()
            val url = webServer.getBaseUrl().resolve("example.git/").resolve(LfsServer.SERVLET_AUTH)
            val storage: LfsHttpStorage = GitAsSvnLfsHttpStorage(url, user!!)

            // Check file is not exists
            Assert.assertNull(storage.getReader(oid, -1))
            storage.getWriter(user).use { writer ->
                writer.write(data)
                Assert.assertEquals(writer.finish(null), oid)
            }
            storage.getWriter(user).use { writer ->
                writer.write(data)
                Assert.assertEquals(writer.finish(null), oid)
            }

            // Read old file.
            val reader = storage.getReader(oid, -1)
            Assert.assertNotNull(reader)
            Assert.assertNull(reader!!.md5)
            Assert.assertEquals(reader.size, data.size.toLong())
            reader.openStream().use { stream ->
                val actual = ByteStreams.toByteArray(stream)
                Assert.assertEquals(actual, data)
            }
            LfsLocalStorageTest.checkLfs(storage, user)
            LfsLocalStorageTest.checkLfs(storage, user)
            LfsLocalStorageTest.checkLocks(storage, user)
        }
    }

    private class GitAsSvnLfsHttpStorage(private val authUrl: URI, private val lfsUser: User) : LfsHttpStorage(), LfsStorageFactory, SharedConfig {
        override fun createStorage(context: LocalContext): LfsStorage {
            return this
        }

        override fun lfsClient(user: User): Client {
            val httpClient = createHttpClient()
            val authProvider: CachedAuthProvider = object : CachedAuthProvider() {
                @Throws(IOException::class)
                override fun getAuthUncached(operation: Operation): Link {
                    val post = HttpPost(authUrl)
                    val params = ArrayList<NameValuePair>()
                    addParameter(params, "secretToken", "t0ken")
                    if (lfsUser.isAnonymous) {
                        addParameter(params, "mode", "anonymous")
                    } else {
                        addParameter(params, "mode", "username")
                        addParameter(params, "userId", lfsUser.username)
                    }
                    post.entity = UrlEncodedFormEntity(params)
                    try {
                        val response: HttpResponse = httpClient.execute(post)
                        if (response.statusLine.statusCode == HttpStatus.SC_OK) {
                            return JsonHelper.mapper.readValue(response.entity.content, Link::class.java)
                        }
                        throw RequestException(post, response)
                    } finally {
                        post.abort()
                    }
                }

                private fun addParameter(params: MutableList<NameValuePair>, key: String, value: String?) {
                    if (value != null) {
                        params.add(BasicNameValuePair(key, value))
                    }
                }
            }
            return Client(authProvider, httpClient)
        }

        override fun create(context: SharedContext) {
            context.add(LfsStorageFactory::class.java, this)
        }

        override fun close() {}
    }
}
